libraries/AP_Winch/AP_Winch.h

```diff
@@ -0,0 +1,89 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_Param/AP_Param.h>
+#include <AC_PID/AC_PID.h>
+#include <AP_WheelEncoder/AP_WheelEncoder.h>
+
+// winch rate control default gains
+#define AP_WINCH_POS_P      1.00f
+#define AP_WINCH_RATE_P     1.00f
+#define AP_WINCH_RATE_I     0.50f
+#define AP_WINCH_RATE_IMAX  1.00f
+#define AP_WINCH_RATE_D     0.00f
+#define AP_WINCH_RATE_FILT  5.00f
+#define AP_WINCH_RATE_DT    0.10f
+
+class AP_Winch_Backend;
+
+class AP_Winch {
+public:
+    AP_Winch();
+
+    // indicate whether this module is enabled or not
+    bool enabled() const { return _enabled; }
+
+    // initialise the winch
+    void init(const AP_WheelEncoder *wheel_encoder = nullptr);
```

Comment:
Is there a reason that the wheel encoder can't actually be a child object of the winch? As outlined it will be harder to have support for multiple winches or use other wheel encoders for future use. Why not create a wheel encoder for each winch backend here, and make it a part of the AP_Winch param table? (Unless wheel encoder has to many nested params to fit like that).

Reply:
@OXINARF has commented on this one.  It crossed my mind but would require some changes to the AP_WheelEncoder class and Rover to allow separating them out (a bit more like AC_PID objects).  I'm not sure which is better or more consistent with what we're doing.  In some ways our move towards singletons is a move away from what's being suggested.